src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -202,68 +209,69 @@ private void IgnoreUnsupportProjectReference(PackageSpec project)
-                   .ToList();
-
-                // get the transitive packages, excluding any already contained in the installed packages
-                List<PackageReference> transitivePackages = packageSpec
-                   .TargetFrameworks
-                   .SelectMany(f => GetTransitivePackageReferencesForFramework(f.FrameworkName, _installedPackages, _transitivePackages, targets))
-                   .GroupBy(p => p.PackageIdentity)
-                   .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
-                   .ToList();
-                return new ProjectPackages(installedPackages, transitivePackages);
+                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<TransitivePackageReference>());
             }
-            else
+
+            IList<LockFileTarget> targetsList = null;
+            if (IsInstalledAndTransitiveComputationNeeded)
             {
-                return new ProjectPackages(Array.Empty<PackageReference>(), Array.Empty<PackageReference>());
+                // clear the transitive packages cache, since we don't know when a dependency has been removed
+                ClearCache();
+                targetsList = await GetTargetsListAsync(token);
             }
+
+            var frameworkSorter = new NuGetFrameworkSorter();
+
+            List<PackageReference> installedPackages = reading
+                .TargetFrameworks
+                .SelectMany(f => GetPackageReferencesForFramework(
+                    f.Dependencies,
+                    f.FrameworkName,
+                    _installedPackages,
+                    targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
+
+            // get the transitive packages, excluding any already contained in the installed packages
+            List<PackageReference> transitivePackages = reading
+                .TargetFrameworks
+                .SelectMany(f => GetTransitivePackageReferencesForFramework(
+                    f.FrameworkName,
+                    _installedPackages,
+                    _transitivePackages,
+                    targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
+
+            List<TransitivePackageReference> transitivePackagesWithOrigins = transitivePackages
+                .Select(transitivePr => Tuple.Create(transitivePr, GetTransitivePackageOrigin(
+                    transitivePr,
+                    installedPackages,
+                    targetsList,
+                    token)))
+                .Select(te => MergeTransitiveOrigin(te.Item1, te.Item2))
+                .Where(transitivePR => transitivePR.TransitiveOrigins.Any())
```

Comment:
@NuGet/nuget-client  Please review here: with this change, we will send only transitive references with a transitive origin.

Transitive packages brought by packages in other projects, brought by a project reference, will not be listed.

Reply:
Can you elaborate why we're doing this? 

Why wouldn't we show all transitive references? Regardless whether they are coming through a project or package.

Reply:
This change just makes testing easier. It seems package recommender uses Transitive list somehow

Reply:
The package recommender uses the transitive list to filter recommendations. We don't want to recommend packages that are already transitively referenced in the project.

Reply:
Given that we're making a functional change, I don't think the  testability argument is a great one. 

We shouldn't exclude transitive packages.

Reply:
Removed filtering.

Reply:
if I understand it correctly, if this filtering did anything, then there's an implementation bug. 

There should be no transitive package with no origin.